### PR TITLE
[TEMPLATES] Rewrite state preparation templates as operations

### PR DESCRIPTION
**Pull request by mariaschuld**
This is the third PR to refactor the templates (following [1156](https://github.com/PennyLaneAI/pennylane/pull/1156) and [1163](https://github.com/PennyLaneAI/pennylane/pull/1163)).

**Changes:**

* Templates are turned into operations, so that the decomposition is executed in the `expand` function.
* The pre-processing was moved into the `__init__` function.
* The tests were moved into a separate file.
* Docstrings were improved.
* If the template takes weights, a method to return the shape of the weights was added.

**Drawbacks:**
The existing  `MottonenStatePreparation` gradient test is insufficient, and it turns out that for most cases gradients _do not_ yet work. However, this is a major task and should be done in a separate PR.

### CODE REVIEW

File: pennylane/templates/state_preparations/mottonen.py [hunk 1]
```diff
@@ -268,36 +229,66 @@ def MottonenStatePreparation(state_vector, wires):
 
         The final state is only equal to the input state vector up to a global phase.
 
+    .. warning::
+
+        Due to non-trivial classical processing of the state vector,
+        this template is not always fully differentiable.
```
Comment (mariaschuld) [inline pennylane/templates/state_preparations/mottonen.py:235]: I added this warning for now.

File: pennylane/tape/tape.py [hunk 2]
```diff
@@ -1103,7 +1103,9 @@ def to_openqasm(self, wires=None, rotations=True):
                     op.inv()
 
         # decompose the queue
-        operations = tape.expand(stop_at=lambda obj: obj.name in OPENQASM_GATES).operations
+        operations = tape.expand(
+            depth=10, stop_at=lambda obj: obj.name in OPENQASM_GATES
```
Comment (mariaschuld) [inline pennylane/tape/tape.py:1107]: We have one more level of "nesting", so need to define depth here. I never know how deep one should make this, but 10 should be more than enough for now

File: tests/templates/test_subroutines.py [hunk 3]
```diff
@@ -1572,7 +1572,7 @@ def test_phase_estimated_two_qubit(self):
                 )
                 qml.probs(estimation_wires)
 
-            tape = tape.expand()
+            tape = tape.expand(depth=2)
```
Comment (mariaschuld) [inline tests/templates/test_subroutines.py:1575]: Same problem as above...

File: tests/tape/test_tape.py [hunk 4]
```diff
@@ -728,7 +729,7 @@ def test_nesting_and_decomposition(self):
             qml.probs(wires=0), qml.probs(wires="a")
 
         new_tape = tape.expand()
-        assert len(new_tape.operations) == 5
+        assert len(new_tape.operations) == 4
```
Comment (mariaschuld) [inline tests/tape/test_tape.py:732]: I thought here we do not have to expand to higher depth, but can just adjust the count...

File: pennylane/tape/tape.py [hunk 5]
```diff
@@ -1103,7 +1103,9 @@ def to_openqasm(self, wires=None, rotations=True):
                     op.inv()
 
         # decompose the queue
-        operations = tape.expand(stop_at=lambda obj: obj.name in OPENQASM_GATES).operations
+        operations = tape.expand(
+            depth=2, stop_at=lambda obj: obj.name in OPENQASM_GATES
```
Comment (mariaschuld) [inline pennylane/tape/tape.py:1107]: We have one more level of nesting, so need to increase the depth here.

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+        ],
+    )
+    def test_shape(self, n_wires, expected_shape):
+        """Test that the shape method returns the correct shape of the weights tensor"""
+
+        shape = qml.templates.ArbitraryStatePreparation.shape(n_wires)
+        assert shape == expected_shape
+
+
+def circuit_template(weights):
```
Comment (albi3ro) [inline tests/templates/test_state_preparations/test_arbitrary_state_prep.py:214]: Should these be fixtures? Just curious.

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+class TestInterfaces:
+    """Tests that the template is compatible with all interfaces, including the computation
+    of gradients."""
+
+    def test_list_and_tuples(self, tol):
+        """Tests common iterables as inputs."""
+
+        weights = [1] * 6
+
+        dev = qml.device("default.qubit", wires=2)
```
Comment (albi3ro) [inline tests/templates/test_state_preparations/test_arbitrary_state_prep.py:238]: Maybe instead take `qubit_device_2_wires` as a fixture input?

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+        res2 = circuit2(weights_tuple)
+        assert qml.math.allclose(res, res2, atol=tol, rtol=0)
+
+    def test_autograd(self, tol):
+        """Tests the autograd interface."""
+
+        weights = np.random.random(size=(6,))
+        weights = pnp.array(weights, requires_grad=True)
+
+        dev = qml.device("default.qubit", wires=2)
```
Comment (albi3ro) [inline tests/templates/test_state_preparations/test_arbitrary_state_prep.py:259]: Same comment as above: should the test instead use the `qubit_device_2_wires` fixture?

Won't continue to leave this comment everywhere applicable.

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+
+        weights_tuple = tuple(weights)
+        res = circuit(weights_tuple)
+        res2 = circuit2(weights_tuple)
+        assert qml.math.allclose(res, res2, atol=tol, rtol=0)
+
+    def test_autograd(self, tol):
+        """Tests the autograd interface."""
+
+        weights = np.random.random(size=(6,))
```
Comment (albi3ro) [inline tests/templates/test_state_preparations/test_arbitrary_state_prep.py:256]: random numbers in tests make me unsettled.  Is there a fixed seed at least?

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+
+        grad_fn2 = qml.grad(circuit2)
+        grads2 = grad_fn2(weights)
+
+        assert np.allclose(grads[0], grads2[0], atol=tol, rtol=0)
+
+    def test_jax(self, tol, skip_if_no_jax_support):
+        """Tests the jax interface."""
+
+        import jax
```
Comment (albi3ro) [inline tests/templates/test_state_preparations/test_arbitrary_state_prep.py:279]:
```suggestion
       jax = pytest.importorskip("jax")
```

We should probably condition the interface tests so people can tests out changes without the full configured environment.

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+
+        grad_fn2 = jax.grad(circuit2)
+        grads2 = grad_fn2(weights)
+
+        assert np.allclose(grads[0], grads2[0], atol=tol, rtol=0)
+
+    def test_tf(self, tol, skip_if_no_tf_support):
+        """Tests the tf interface."""
+
+        import tensorflow as tf
```
Comment (albi3ro) [inline tests/templates/test_state_preparations/test_arbitrary_state_prep.py:304]:
```suggestion
        tf = pytest.importorskip("tensorflow")
```

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+        with tf.GradientTape() as tape2:
+            res2 = circuit2(weights)
+        grads2 = tape2.gradient(res2, [weights])
+
+        assert np.allclose(grads[0], grads2[0], atol=tol, rtol=0)
+
+    def test_torch(self, tol, skip_if_no_torch_support):
+        """Tests the torch interface."""
+
+        import torch
```
Comment (albi3ro) [inline tests/templates/test_state_preparations/test_arbitrary_state_prep.py:330]:
```suggestion
        torch = pytest.importorskip("torch")
```

File: tests/templates/test_state_preparations/test_basis_state_prep.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    """Test inputs and pre-processing."""
+
+    def test_exception_wrong_dim(self):
+        """Verifies that exception is raised if the
+        number of dimensions of features is incorrect."""
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        def circuit(weights):
+            qml.templates.ArbitraryStatePreparation(weights, wires=range(3))
```
Comment (albi3ro) [inline tests/templates/test_state_preparations/test_basis_state_prep.py:112]: I believe this test is in the wrong file.

File: pennylane/templates/state_preparations/mottonen.py [hunk 14]
```diff
@@ -268,36 +229,66 @@ def MottonenStatePreparation(state_vector, wires):
...
+
+        n_amplitudes = shape[0]
+        if n_amplitudes != 2 ** len(wires):
+            raise ValueError(
+                f"State vector must be of length {2 ** len(wires)} or less; got length {n_amplitudes}."
+            )
+
+        norm = qml.math.sum(qml.math.abs(state_vector) ** 2)
+        if not qml.math.allclose(norm, 1.0, atol=1e-3):
+            raise ValueError("State vector has to be of length 1.0, got {}".format(norm))
```
Comment (albi3ro) [inline pennylane/templates/state_preparations/mottonen.py:263]:
```suggestion
            raise ValueError("State vector has to be of norm 1.0, got {}".format(norm))
```

Two different types of errors are "State vector must be of length ...". The two different uses of the word length are confusing.  Wrong size of `state_vector` should use a different word than wrong normalization of `state_vector`.

**Review by albi3ro — CHANGES_REQUESTED** [review]
Do we need to catch if state preparation is not occurring at the beginning of a circuit on a 00000... state?

I tested
```
@qml.qnode(dev)
def circuit():
    qml.PauliX(wires=(0))
    qml.templates.state_preparations.basis.BasisStatePreparation([1,0,0], wires=(0,1,2))
    return qml.state()
```
and that lead to non-intuitive behaviour.

I left a lot of comments, but most of them are just suggestions.

The things that should really be changed are:
* placing `test_basis_state_prep.TestInputs.test_exception_wrong_dim` in the correct file
* using `pytest.importorskip` in interface tests
* docstring for `ArbitraryStatePreparation`

File: pennylane/templates/state_preparations/arbitrary_state_preparation.py [hunk 15]
```diff
@@ -79,19 +62,53 @@ def ArbitraryStatePreparation(weights, wires):
...
+                f"Weights tensor must be of shape {(2 ** (len(wires) + 1) - 2,)}; got {shape}."
+            )
+
+        super().__init__(weights, wires=wires, do_queue=do_queue)
+
+    def expand(self):
+
+        with qml.tape.QuantumTape() as tape:
+            for i, pauli_word in enumerate(_state_preparation_pauli_words(len(self.wires))):
+                qml.PauliRot(self.parameters[0][i], pauli_word, wires=self.wires)
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/templates/state_preparations/arbitrary_state_preparation.py:100]: The parameters of an `Operation` are by definition stored in a list. Even if we only have one parameter array, it will be the first element of the parameter list. 

I was wondering if it is clearer to define 
```
weights = self.parameters[0]
``` 
which would help coders?

File: pennylane/templates/state_preparations/arbitrary_state_preparation.py [hunk 16]
```diff
@@ -79,19 +62,53 @@ def ArbitraryStatePreparation(weights, wires):
...
+            raise ValueError(
+                f"Weights tensor must be of shape {(2 ** (len(wires) + 1) - 2,)}; got {shape}."
+            )
+
+        super().__init__(weights, wires=wires, do_queue=do_queue)
+
+    def expand(self):
+
+        with qml.tape.QuantumTape() as tape:
+            for i, pauli_word in enumerate(_state_preparation_pauli_words(len(self.wires))):
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/templates/state_preparations/arbitrary_state_preparation.py:99]: There is actually a reason I am trying to use indices: Indexing always works nicely through all interfaces. But iterating over tensors does not, for example there were issues with tf tensors that are not iterables. Also I suspect that indexing is faster.

I have to admit it would be good to systematically check this, but for now I am trying to stick to iterating over indices...

File: pennylane/templates/state_preparations/basis.py [hunk 17]
```diff
@@ -12,71 +12,54 @@
...
+            raise ValueError(f"Basis state must be of length {len(wires)}; got length {n_bits}.")
 
+        # we can extract a list here, because embedding is not differentiable
+        basis_state = list(qml.math.toarray(basis_state))
 
-@template
-def BasisStatePreparation(basis_state, wires):
-    r"""
-    Prepares a basis state on the given wires using a sequence of Pauli X gates.
+        if not all(bit in [0, 1] for bit in basis_state):
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/templates/state_preparations/basis.py:54]: I was trying to only clean up the worst bits, since the PR is already so large...but can add an example!

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py (hunk 6 above)
Comment (mariaschuld) [reply tests/templates/test_state_preparations/test_arbitrary_state_prep.py:214]: I think it's ok if they are just functions defined somewhere...since they change in every test, so it would be no use to make them global fixtures...Or did I understand wrong?

I am never sure if I should put them into the test class, but I find it so ugly to refer to them with `self`...

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py (hunk 7 above)
Comment (mariaschuld) [reply tests/templates/test_state_preparations/test_arbitrary_state_prep.py:238]: I think our current idea is to move away from fixtures, to make the tests as self-contained as possible. Ideally we should be able just to copy-and-paste the test and it runs!

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py (hunk 9 above)
Comment (mariaschuld) [reply tests/templates/test_state_preparations/test_arbitrary_state_prep.py:256]: I think here it is fine, because the test should really work for _any_ random number? I can fix a seed, but it feels pointless then to use random numbers at all?

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py (hunk 10 above)
Comment (mariaschuld) [reply tests/templates/test_state_preparations/test_arbitrary_state_prep.py:279]: This is taken care of by the fixture `skip_if_no_jax_support`! I also find it strange but it works by just passing it to the test :) 

Do you think it is safer to do the "importorskip" in addition?

Comment (mariaschuld) [reply tests/templates/test_state_preparations/test_arbitrary_state_prep.py:279]: Good point. I think I misunderstood the `skip_if_no_jax_support` fixture. But in the interest of being self-contained, it is better to just apply it directly!

File: tests/templates/test_state_preparations/test_basis_state_prep.py (hunk 13 above)
Comment (mariaschuld) [reply tests/templates/test_state_preparations/test_basis_state_prep.py:112]: Ah, good find!

File: pennylane/templates/state_preparations/mottonen.py (hunk 14 above)
Comment (mariaschuld) [reply pennylane/templates/state_preparations/mottonen.py:263]: Good point!

**Comment by mariaschuld** [comment]
Thanks @albi3ro for your great review!

The bug about `BasisStatePreparation` not being checked for being the first operation in a circuit is indeed confusing. Can we solve this in a separate issue though, since it would be a breaking change? I will open one!

File: pennylane/templates/state_preparations/arbitrary_state_preparation.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/templates/state_preparations/arbitrary_state_preparation.py:99]: Thanks for the explanation :)

So I just ran a quick check, and it seems like using zip is *slightly* faster. but it's more important to make sure it works in all interfaces.

File: pennylane/templates/state_preparations/arbitrary_state_preparation.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/templates/state_preparations/arbitrary_state_preparation.py:100]: I think it's fine the way it is.  Most of the time I would just accept a detail like that, but I was being particular and curious.

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py (hunk 7 above)
Comment (albi3ro) [reply tests/templates/test_state_preparations/test_arbitrary_state_prep.py:238]: Good to know

File: tests/templates/test_state_preparations/test_arbitrary_state_prep.py (hunk 9 above)
Comment (albi3ro) [reply tests/templates/test_state_preparations/test_arbitrary_state_prep.py:256]: The problem lies when something goes wrong and the test is failing.  Then the test and the failure isn't necessarily reproducible for easy debugging.

But this may be another change outside the scope of the PR, and *shouldn't* be too problematic here.

```suggestion
        weights = np.array([0.11, 1.234, 2.0, -0.590, -3.001, 0.003])
```

**Review by albi3ro — APPROVED** [review]
Looks great!

It was great to learn about a new section of code and thanks for answering all my questions and nitpicks 👍
